discovery/kubernetes/kubernetes.go

```diff
@@ -101,19 +122,91 @@ func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if err != nil {
 		return err
 	}
-	if c.Role == "" {
-		return errors.Errorf("role missing (one of: pod, service, endpoints, node, ingress)")
-	}
 	err = c.HTTPClientConfig.Validate()
 	if err != nil {
 		return err
 	}
 	if c.APIServer.URL == nil && !reflect.DeepEqual(c.HTTPClientConfig, config_util.HTTPClientConfig{}) {
 		return errors.Errorf("to use custom HTTP client configuration please provide the 'api_server' URL explicitly")
 	}
+	switch c.Role {
+	case "pod":
+		for _, selector := range c.Selectors {
+			if selector.Role != RolePod && selector.Field != "" {
+				return errors.Errorf("pod role supports only pod selectors")
+			}
+
+			err = validateSelectors(selector)
+			if err != nil {
+				return err
+			}
+		}
+	case "service":
+		for _, selector := range c.Selectors {
+			if selector.Role != RoleService && selector.Field != "" {
+				return errors.Errorf("service role supports only service selectors")
+			}
+
+			err = validateSelectors(selector)
+			if err != nil {
+				return err
+			}
+		}
+	case "endpoints":
+		for _, selector := range c.Selectors {
+			if selector.Role != RoleEndpoint && selector.Role != RolePod && selector.Role != RoleService && selector.Field != "" {
+				return errors.Errorf("endpoints role supports only pod, service and endpoints selectors")
+			}
+
+			err = validateSelectors(selector)
+			if err != nil {
+				return err
+			}
+		}
+	case "node":
+		for _, selector := range c.Selectors {
+			if selector.Role != RoleNode && selector.Field != "" {
+				return errors.Errorf("node role supports only node selectors")
+			}
+
+			err = validateSelectors(selector)
+			if err != nil {
+				return err
+			}
+		}
+	case "ingress":
+		for _, selector := range c.Selectors {
+			if selector.Role != RoleIngress && selector.Field != "" {
+				return errors.Errorf("ingress role supports only ingress selectors")
+			}
+
+			err = validateSelectors(selector)
+			if err != nil {
+				return err
+			}
+		}
+	default:
+		return errors.Errorf("role missing (one of: pod, service, endpoints, node, ingress)")
+	}
+	foundSelectorRoles := make(map[Role]bool)
```

Comment:
(nit) I would use a map[Role]struct{} for `foundSelectorRoles` as it's more commonly used for detecting duplicates.
Also maybe you could validate selectors in the `for _, selector := range c.Selectors` loop instead of repeating the same code in each `switch` block?